pkg/logql/log/fmt.go

```diff
@@ -407,6 +407,7 @@ func (lf *LabelsFormatter) Process(ts int64, l []byte, lbs *LabelsBuilder) ([]by
 			lbs.SetErrorDetails(err.Error())
 			continue
 		}
+		lbs.Del(f.Name) // Deleting to avoid duplication
```

Comment:
Thank you @ravishankar15 for creating the PR!

> @salvacorts Maybe I am overly cautious and my worries are ungrounded, so I want your opinion on this.

I agree @chaudum, this problem may not only limited to the label formatter, but to any pipeline expression that may overwrite a label that already exists in structured metadata.

The bug is in the [`LabelsBuilder.UnsortedLabels`](https://github.com/grafana/loki/blob/4c88be0ef2310406206c02d7747bf73b51c95bda/pkg/logql/log/labels.go#L406) method, which seems to be used only by `IntoMap` (later used by `LabelsFormatter.Process`). This is what I'd do:

- Similarly to [how we skip stream labels that will be replaced][1], skip structured metadata fields that will be replaced.
- The [`BaseLabelsBuilder.add` field][2] is an array of 3 items (one per category). Each item is a list of labels to add for the category. I would replace this by two new fields:
    - `addStructuredMetadata` and `addParsed`
    - Then in  LabelsBuilder.UnsortedLabels, we can filter out stream labels looking at `addStructuredMetadata` and `addParsed`, and then filter out structured metadata looking at the `addParsed`

On top of this, I'd add tests to the [labels test file](pkg/logql/log/labels_test.go)

How confortable would you feel making these changes @ravishankar15? We are happy to help!

[2]: https://github.com/grafana/loki/blob/4c88be0ef2310406206c02d7747bf73b51c95bda/pkg/logql/log/labels.go#L130
[1]: https://github.com/grafana/loki/blob/4c88be0ef2310406206c02d7747bf73b51c95bda/pkg/logql/log/labels.go#L433-L452

Reply:
Thanks for taking a look at the PR will check these out over the weekend

Reply:
As @chaudum  mentioned I think there are two parts of the problem.

1. The suggestion from Salvacorts will solve the problem where we will not have duplicate values in the response
2. For `label filter after a label format expression needs to match the newly assigned value.` as mentioned by Chandum we need to make changes to the `getWithCategory` method on that note we are setting the flag `referencedStructuredMetadata` to true if the label is from `StructuredMetadata` in our issue the label is from `StructuredMetadata` but its renamed by LabelFormat in `ParsedLabel` so if we make the change the flag will be false I hope that should be okay and expected behaviour.
3. Also I believe we should handle the case in `Add` and `Set` method of LabelsBuilder ? 

I believe we need to solve both the issues for this bug to be resolved truly. Can you clarify my (2) and (3) and I can change my PR accordingly. 

Also, In @salvacorts suggestion was to move away from having single field `BaseLabelsBuilder.add` implementation and have separate fields `addStructuredMetadata` and `addParsed`. Do you feel we need to combine both the changes in a single PR. I feel it will be little cumbersome If its okay shall we split it into two PR's like

1. Fixes the two issue mentioned in chandum comment -> With existing `BaseLabelsBuilder.add`  implementation
2. Modify the `BaseLabelsBuilder.add` implementation to use separate fields `addStructuredMetadata` and `addParsed`

Reply:
I have made the changes to the builder as per our discussion without reformatting which I am hoping should fix the issue.

Also @salvacorts I see the StreamLabels is not used logically outside of builder but I see some tests using that category to add labels and test some scenarios are all those outdated ?  Let me know I can raise a separate PR as a follow up to remove all those stale logic and clean up the builder.